File: go/producer.go
// producer.go
package main

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
)

// MyMessage — пример структуры, которую будем сериализовать в JSON.
type MyMessage struct {
	MsgID int    `json:"msg_id"`
	Text  string `json:"text"`
}

func main() {
	// Конфиг продюсера. acks=all для «at least once».
	producer, err := kafka.NewProducer(&kafka.ConfigMap{
		"bootstrap.servers": "127.0.0.1:9094,127.0.0.1:9095,127.0.0.1:9096",
		"acks":              "all",
	})
	if err != nil {
		panic(err)
	}
	defer producer.Close()

	// Goroutine для чтения результатов доставки (Delivery Reports).
	go func() {
		for e := range producer.Events() {
			switch ev := e.(type) {
			case *kafka.Message:
				if ev.TopicPartition.Error != nil {
					fmt.Printf("Delivery failed: %v\n", ev.TopicPartition.Error)
				} else {
					fmt.Printf("Message delivered to %v [%d] at offset %v\n",
						*ev.TopicPartition.Topic, ev.TopicPartition.Partition, ev.TopicPartition.Offset)
				}
			}
		}
	}()

	// Отправляем 30 тестовых сообщений
	topic := "my_topic"
	for i := 0; i < 30; i++ {
		// Подготовим объект сообщения
		msgObj := MyMessage{
			MsgID: i,
			Text:  fmt.Sprintf("Hello from Go producer, index=%d", i),
		}

		// Сериализация в JSON
		valBytes, err := json.Marshal(msgObj)
		if err != nil {
			fmt.Printf("[Producer] Error while serializing message: %v\n", err)
			continue
		}

		fmt.Printf("Sending: %+v\n", msgObj)

		// Отправляем в Kafka
		err = producer.Produce(&kafka.Message{
			TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
			Value:          valBytes,
		}, nil)
		if err != nil {
			fmt.Printf("[Producer] Produce error: %v\n", err)
		}

		time.Sleep(1 * time.Second)
	}

	// Ждём подтверждения отправки всех сообщений (макс. 15 секунд)
	producer.Flush(15 * 1000)

	fmt.Println("Producer finished sending messages.")
}

File: python/lab1/producer.py
# producer.py
import json
import time
from confluent_kafka import Producer

class MyMessage:
    """Простейший класс для примера сериализации."""
    def __init__(self, msg_id, text):
        self.msg_id = msg_id
        self.text = text

def main():
    # Настройки продюсера
    # "acks=all" даёт гарантию "at least once" при условии, что консюмер корректно коммитит оффсеты.
    producer_conf = {
        'bootstrap.servers': '127.0.0.1:9094,127.0.0.1:9095,127.0.0.1:9096',
        'acks': 'all',
    }
    producer = Producer(producer_conf)

    def delivery_report(err, msg):
        """Callback, вызывается при подтверждении (или ошибке) доставки сообщения брокером."""
        if err is not None:
            print(f"Delivery failed for message {msg.key()}: {err}")
        else:
            print(f"Message delivered to {msg.topic()} [{msg.partition()}] @ offset {msg.offset()}")

    # Отправим тестовые сообщения
    for i in range(30):
        # Генерируем объект сообщения
        message_object = MyMessage(i, f"Hello from python producer, index={i}")
        try:
            # Сериализуем в JSON
            serialized_value = json.dumps(message_object.__dict__).encode('utf-8')
            # Логируем, что отправляем
            print(f"Sending: {message_object.__dict__}")

            producer.produce(
                topic='my_topic',
                value=serialized_value,
                on_delivery=delivery_report
            )
        except Exception as e:
            print(f"[Producer] Error while serializing or sending message: {e}")

        time.sleep(1)

    # Ждём окончания отправки всех сообщений
    producer.flush()

if __name__ == '__main__':
    main()
